jf_agent/main.py

```diff
@@ -86,106 +127,214 @@ def main():
     if pull_since:
         pull_since = pull_since.replace(tzinfo=pytz.utc)
 
-    pull_until = dateparser.parse(args.until) if args.until else datetime.utcnow()
+    now = datetime.utcnow()
+    pull_until = dateparser.parse(args.until) if args.until else now
     pull_until = pull_until.replace(tzinfo=pytz.utc)
 
     if not jira_url and not git_url:
-        print('ERROR: Config file must provide either a Jira or Bitbucket URL.')
+        print('ERROR: Config file must provide either a Jira or Git URL.')
         parser.print_help()
-        sys.exit(1)
+        return
 
     if skip_ssl_verification:
         print('WARNING: Disabling SSL certificate validation')
         # To silence "Unverified HTTPS request is being made."
         urllib3.disable_warnings()
 
+    if run_mode_includes_download:
+        if args.prev_output_dir:
+            print('ERROR: Provide output_basedir if downloading, not prev_output_dir')
+            return
+
+        output_basedir = args.output_basedir
+        output_dir_timestamp = now.strftime('%Y%m%d_%H%M%S')
+        outdir = os.path.join(output_basedir, output_dir_timestamp)
+        try:
+            os.makedirs(outdir, exist_ok=False)
+        except FileExistsError:
+            print(f"ERROR: Output dir {outdir} already exists")
+            return
+        except Exception:
+            print(
+                f"ERROR: Couldn't create output dir {outdir} -- bad OUTPUT_BASEDIR ({output_basedir})?"
+            )
+            return
+        print(f'Will write output files into {outdir}')
+
+    if run_mode_is_print_all_jira_fields and not jira_url:
+        print(f'ERROR: Must provide jira_url for mode {run_mode}')
+
+    if run_mode_includes_send and not run_mode_includes_download:
+        if not args.prev_output_dir:
+            print('ERROR: prev_output_dir must be provided if not downloading')
+            return
+        if not os.path.isdir(args.prev_output_dir):
+            print(f'ERROR: prev_output_dir ("{args.prev_output_dir}") is not a directory')
+            return
+        outdir = args.prev_output_dir
+
+    api_token = os.environ.get('JELLYFISH_API_TOKEN')
+    if not api_token:
+        print(f'ERROR: JELLYFISH_API_TOKEN not found in the environment.')
+        return
+
+    resp = requests.get(f'{JELLYFISH_API_BASE}/agent/config?api_token={api_token}')
+    if not resp.ok:
+        print(
+            f"ERROR: Couldn't get agent config info from {JELLYFISH_API_BASE}/agent/config "
+            f'using provided JELLYFISH_API_TOKEN (HTTP {resp.status_code})'
+        )
+        return
```

Comment:
nit: I think it would be better to use `sys.exit(1)` here (and other places where we `return` based on bad configuration). In cases where the agent is being run non-interactively (e.g. systemd timer unit, cron job) we would want to signal that something has gone wrong. Specifically in the API failure case (which could be transient) we would want to make sure we weren't silently failing.

Reply:
Yeah agreed we should do more to support meaningful exit codes. Previously it had been existing with status 1 in some but not all error scenarios. I started down a path of standardizing, but punted on it realizing that there were nuances that weren't necessarily clear calls (e.g., partial successes). I'll ticket out taking a harder look at this.